Workaround `[AllowShared] BufferSource`

Constellation opened the pull request:
#### 02a7ddd1949f734535c90d93a6d10eea06afcdce
<pre>
Workaround `[AllowShared] BufferSource`
<a href="https://bugs.webkit.org/show_bug.cgi?id=249632">https://bugs.webkit.org/show_bug.cgi?id=249632</a>
rdar://103543109

Reviewed by Mark Lam.

BufferSource is typedef ArrayBuffer or ArrayBufferView. However, WebCore is implementing it as one class instead
of using Variant. As a result IDL AllowShared annotation is not applied correctly to each of variant&apos;s type.
We already hit this issue in WebGL and we workaround this by using ([AllowShared] ArrayBufferView or [AllowShared] ArrayBuffer).
We should do that in all the other IDL files too.

* Source/WebCore/Modules/WebGPU/GPUQueue.idl:
* Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.idl:
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.idl:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoderConfig.idl:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.idl:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(AddToIncludesForIDLType):
* Source/WebCore/dom/TextDecoder.idl:
* Source/WebCore/dom/TextDecoderStreamDecoder.idl:

Canonical link: <a href="https://commits.webkit.org/258137@main">https://commits.webkit.org/258137@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4dddc45bb74a0d8cee3da745f637f96ae4867c25

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100998 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10154 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34051 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110302 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170557 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104985 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11084 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1026 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93406 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108136 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106781 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8388 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91650 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34999 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90296 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23047 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77990 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3823 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24564 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3851 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/966 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9963 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44072 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5619 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2930 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/WebCore/Modules/WebGPU/GPUQueue.idl:
```diff
@@ -44,13 +44,13 @@ interface GPUQueue {
         GPUBuffer buffer,
         GPUSize64 bufferOffset,
-        [AllowShared] BufferSource data,
+        [AllowShared] (ArrayBufferView or ArrayBuffer) data,
```
In `WebGLRenderingContextBase.idl`, I see:
```
typedef ([AllowShared] ArrayBuffer or [AllowShared] ArrayBufferView) AllowSharedBufferDataSource;
typedef ([AllowShared] Float32Array or sequence<GLfloat>) Float32List;
```
... where the parentheses goes around the whole thing.  Is it OK that your parentheses goes around the 2 variant types?

Constellation replied:
Yes, this is fine, attaching `[AllowShared]` to `(ArrayBufferView or ArrayBuffer)` union.

MenloDorian reviewed (approved):
r=me

Constellation committed:
Workaround `[AllowShared] BufferSource`
https://bugs.webkit.org/show_bug.cgi?id=249632
rdar://103543109

Reviewed by Mark Lam.

BufferSource is typedef ArrayBuffer or ArrayBufferView. However, WebCore is implementing it as one class instead
of using Variant. As a result IDL AllowShared annotation is not applied correctly to each of variant's type.
We already hit this issue in WebGL and we workaround this by using ([AllowShared] ArrayBufferView or [AllowShared] ArrayBuffer).
We should do that in all the other IDL files too.

* Source/WebCore/Modules/WebGPU/GPUQueue.idl:
* Source/WebCore/Modules/filesystemaccess/FileSystemSyncAccessHandle.idl:
* Source/WebCore/Modules/webcodecs/WebCodecsEncodedVideoChunk.idl:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoDecoderConfig.idl:
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.idl:
* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
(AddToIncludesForIDLType):
* Source/WebCore/dom/TextDecoder.idl:
* Source/WebCore/dom/TextDecoderStreamDecoder.idl:

Canonical link: https://commits.webkit.org/258137@main
